mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -0,0 +1,470 @@
+    CallOpInterface Call = It.Call;
+
+    bool DoInline = Inliner.shouldInline(It);
+    LLVM_DEBUG(llvm::dbgs() << ((DoInline) ? "* Inlining call: "
+                                           : "* Not inlining call: ")
+                            << I << ". " << Call << "\n";);
+    if (!DoInline)
+      continue;
+
+    Region *TgtRegion = It.TgtNode->getCallableRegion();
+    LogicalResult InlineRes = inlineCall(
+        Inliner, Call, cast<CallableOpInterface>(TgtRegion->getParentOp()),
+        TgtRegion, /*shouldCloneInlinedRegion=*/true);
+
+    if (failed(InlineRes)) {
+      LLVM_DEBUG(llvm::dbgs() << "** Failed to inline\n");
+      continue;
+    }
+
+    LLVM_DEBUG(llvm::dbgs() << "** Inline succeeded\n");
+
+    // then erase the call.
+    Call.erase();
+  }
+
+  Inliner.clear();
+}
+
+void Inliner::collectCallOps(iterator_range<Region::iterator> Blocks,
+                             CallGraphNode *SrcNode, CallGraph &CG,
+                             SymbolTableCollection &SymTable,
+                             SmallVectorImpl<ResolvedCall> &Calls,
+                             bool TraverseNestedCgNodes) {
+  SmallVector<std::pair<Block *, CallGraphNode *>, 8> WorkList;
+  auto AddToWorkList = [&](CallGraphNode *CGN,
+                           iterator_range<Region::iterator> Blocks) {
+    for (Block &B : Blocks)
+      WorkList.emplace_back(&B, CGN);
+  };
+
+  AddToWorkList(SrcNode, Blocks);
+
+  while (!WorkList.empty()) {
+    Block *B;
+    std::tie(B, SrcNode) = WorkList.pop_back_val();
+
+    for (Operation &Op : *B) {
+      if (auto Call = dyn_cast<CallOpInterface>(Op)) {
+        CallInterfaceCallable Callable = Call.getCallableForCallee();
+        if (SymbolRefAttr SymRef = dyn_cast<SymbolRefAttr>(Callable)) {
+          if (!SymRef.isa<FlatSymbolRefAttr>())
+            continue;
+        }
+
+        CallGraphNode *TgtNode = CG.resolveCallable(Call, SymTable);
+        if (!TgtNode->isExternal())
+          Calls.emplace_back(Call, SrcNode, TgtNode);
+        continue;
+      }
+
+      // If this is not a call, traverse the nested regions. If
+      // `TraverseNestedCGNodes` is false, then don't traverse nested call graph
+      // regions.
+      for (auto &NestedRegion : Op.getRegions()) {
+        CallGraphNode *NestedNode = CG.lookupNode(&NestedRegion);
+        if (TraverseNestedCgNodes || !NestedNode)
+          AddToWorkList(NestedNode ? NestedNode : SrcNode, NestedRegion);
+      }
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// AlwaysInliner
+//===----------------------------------------------------------------------===//
+
+/// Returns true if the given call should be inlined.
+bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
+  // Don't allow inlining terminator calls. We currently don't support this
+  // case.
+  if (ResolvedCall.Call->hasTrait<OpTrait::IsTerminator>())
+    return false;
```

Comment:
Is this possible for our calls?

Reply:
No, but we are likely to want to inline func.call too and I am not sure if that can be a terminator.